src/main/java/me/kavin/piped/server/handlers/auth/UserHandlers.java

```diff
@@ -109,10 +132,277 @@ public static byte[] loginResponse(String user, String pass)
+        )
+                .endpointURI(provider.authUri)
+                .codeChallenge(pkceVerifier, CodeChallengeMethod.S256)
+                .state(new State(state))
+                .nonce(data.getOidNonce());
+
+                if (provider.sendMaxAge) {
+                    // This parameter is optional and the idp doesn't have to honor it.
+                    oidcRequestBuilder.maxAge(0);
+                }
+
+        return HttpResponse.redirect302(oidcRequestBuilder.build().toURI().toString());
+    }
+
+    public static HttpResponse oidcDeleteCallback(OidcProvider provider, URI requestUri) throws Exception {
+
+        AuthenticationSuccessResponse sr = parseOidcUri(requestUri);
+
+        OidcData data = DatabaseHelper.getOidcData(sr.getState().toString());
+
+        if (data == null) {
+            return HttpResponse.ofCode(400).withHtml(
+                    "Your oidc provider sent invalid state data. Try again or contact your oidc admin"
+            );
+        }
+
+        String redirect = data.data.split("\\|")[1];
+        String session = data.data.split("\\|")[0];
+
+        URI callback = new URI(Constants.PUBLIC_URL + "/oidc/" + provider.name + "/delete");
+        AuthorizationCode code = sr.getAuthorizationCode();
+
+        if (code == null) {
+            return HttpResponse.ofCode(400).withHtml(
+                    "Your oidc provider sent an invalid code. Try again or contact your oidc admin"
+            );
+        }
+
+        AuthorizationGrant codeGrant = new AuthorizationCodeGrant(code, callback, data.getOidVerifier());
+
+        ClientAuthentication clientAuth = new ClientSecretBasic(provider.clientID, provider.clientSecret);
+
+        TokenRequest tokenRequest = new TokenRequest.Builder(provider.tokenUri, clientAuth, codeGrant).build();
+        TokenResponse tokenResponse = OIDCTokenResponseParser.parse(tokenRequest.toHTTPRequest().send());
+
+        if (!tokenResponse.indicatesSuccess()) {
+            TokenErrorResponse errorResponse = tokenResponse.toErrorResponse();
+            return HttpResponse.ofCode(500).withHtml("Failure while trying to request token:\n\n" + errorResponse.getErrorObject().getDescription());
+        }
+
+        OIDCTokenResponse successResponse = (OIDCTokenResponse) tokenResponse.toSuccessResponse();
+
+        JWT idToken = JWTParser.parse(successResponse.getOIDCTokens().getIDTokenString());
+
+        IDTokenClaimsSet claims;
+        try {
+            claims = provider.validator.validate(idToken, data.getOidNonce());
+        } catch (BadJOSEException e) {
+            System.err.println("Invalid token received: " + e);
+            return HttpResponse.ofCode(400).withHtml("Received a bad token. Please try again");
+        } catch (JOSEException e) {
+            System.err.println("Token processing error: " + e);
+            return HttpResponse.ofCode(500).withHtml("Internal processing error. Please try again");
+        }
```

Comment:
That's just a lot of duplicate code, it would certainly make sense to refactor the common code into a new method.